### DISCUSSION TITLE
Implementation MCCS via USB (EIZO)

### DISCUSSION THREAD

**Discussion by gustavneustadt** [MonitorControl/MonitorControl#717] [General]
I have a EV2785 and sadly MonitorControl doesnt work for me. I had contact with EIZO support and they confirmed to me that they do not build DDC/CI controllers in their monitors anymore. They seem to use their own proprietary windows software. (#351)

I found [dccutil](https://github.com/rockowitz/ddcutil/) supports the MCCS protocol which is necessary (?) for monitors from Eizo. It communicates not via hdmi/dp but via usb with the monitors to control them. I didnt tried it, because I was not able to build it on my mac.

Would be nice to see a implementation of this.

**Comment by waydabber** [comment]
You are right, this could be a useful feature. I'll try to add this in the future (but no promises).

"Alternatively, some monitors (e.g. Eizo ColorEdge, Apple Cinema) provide a USB interface to the VCP, as described in the USB Monitor Control Class Specification. ddcutil can communicate with these monitors over USB instead of I2C."

**Comment by waydabber** [comment]
For future reference - ddcctl seems to use this to detect such displays:

https://github.com/rockowitz/ddcutil/blob/39ce66b42305b304816d635b42a5b38ce280e2dd/src/usb/usb_displays.c#L693-L743

Also, for Apple branded external displays standard DisplayService calls should work (this is now limited to the internal display only). The following could be used to detect such displays using a whitelist method and direct them to use DisplayService instead of DDC:

let ULTRAFINE_NAME = "LG UltraFine"
let THUNDERBOLT_NAME = "Thunderbolt"
let LED_CINEMA_NAME = "LED Cinema"
let CINEMA_NAME = "Cinema"
let CINEMA_HD_NAME = "Cinema HD"
let COLOR_LCD_NAME = "Color LCD"
let APPLE_DISPLAY_VENDOR_ID = 0x05AC

(credits for this goes to @alin23)

@gustavneustadt - a question: does the built in MacOS brightness slider control the Eizo display? Or even though it implements a similar protocol to Apple's Cinema display it is not being controlled by the OS? Thank you!

**Comment by alin23** [comment]
One can easily test if DisplayServices works for their monitor using the Lunar CLI.

1. Install the CLI from the menu bar as seen here: https://lunar.fyi/#cli
2. Get the Serial/UUID of the display you want to test using this command:
```sh
lunar displays | egrep '^\s+Name:|Serial:'
```
3. And then test `DisplayServicesSetBrightness` and `CoreDisplay_Display_SetUserBrightness` using the following commands (run them one by one to see which one works)

##### Set brightness `20` for UUID `37D8832A-2D66-02CA-B9F7-8F30A301B230` using `DisplayServices`
```sh
lunar display-services SetBrightness 37D8832A-2D66-02CA-B9F7-8F30A301B230 20
```

##### Set brightness to `20` for UUID `37D8832A-2D66-02CA-B9F7-8F30A301B230` using `CoreDisplay`
```sh
lunar core-display SetUserBrightness 37D8832A-2D66-02CA-B9F7-8F30A301B230 20
```

**Comment by waydabber** [comment]
Thanks @alin23 !

@gustavneustadt - please let us know if you have the results. Thank you!

**Comment by gustavneustadt** [comment]
Thank you very much! I am currently on vacation but will try it for sure.

UPDATE:

@waydabber

> does the built in MacOS brightness slider control the Eizo display? Or even though it implements a similar protocol to Apple's Cinema display it is not being controlled by the OS? Thank you!

Which one? Didnt find any software slider. If you mean the keyboard controls: No, this do not work. 

@alin23

> One can easily test if DisplayServices works for their monitor using the Lunar CLI.
> 
>     Install the CLI from the menu bar as seen here: https://lunar.fyi/#cli
>     Get the Serial/UUID of the display you want to test using this command:
> 
> lunar displays | egrep '^\s+Name:|Serial:'
> 
>     And then test DisplayServicesSetBrightness and CoreDisplay_Display_SetUserBrightness using the following commands (run them one by one to see which one works)
> 
> Set brightness 20 for UUID 37D8832A-2D66-02CA-B9F7-8F30A301B230 using DisplayServices
> 
> lunar display-services SetBrightness 37D8832A-2D66-02CA-B9F7-8F30A301B230 20
> 
> Set brightness to 20 for UUID 37D8832A-2D66-02CA-B9F7-8F30A301B230 using CoreDisplay
> 
> lunar core-display SetUserBrightness 37D8832A-2D66-02CA-B9F7-8F30A301B230 20

Sadly this does not work for me.
Since I am using a Hackintosh, the problem may come from faulty usb mapping or other hacky stuff – at the moment I have no other way to test this.

**Comment by deg0nz** [comment]
Hi,

just a qick feedback:

I was able to test the Lunar stuff described by @alin23 with a MacBook and an Eizo EV2785 connected via USB-C. Lunar is NOT able to change the brightness of the display.

**Comment by waydabber** [comment]
@gustavneustadt,

I meant the Dispaly Brightness slider in Big Sur's control center. But if the keyboards don't work, then probably EIZO displays are not regarded the same way as Apple displays even though they use a similar protocol.

<img width="338" alt="Screen Shot 2021-08-25 at 18 41 11" src="https://user-
images.githubusercontent.com/37590873/130830955-34e6499d-9f49-45b1-9f8a-ff94396e7987.png">

@deg0nz,

thanks for the feedback.

We should probably divide this issue into two - supporting Apple displays is easy by using DisplayServices while supporting MCCS via USB is an entirely different thing to do.

**Comment by waydabber** [comment]
Apple branded displays will be handled here:

https://github.com/MonitorControl/MonitorControl/issues/539

**Comment by deg0nz** [comment]
One more update:

I digged into this a little bit, compiled `ddcutil` in a Linux VM and tried if I could change/access my monitor's settings by redirecting the USB device of the monitor into the VM.

Unfortunately, it didn't work. 

I also stumbled across https://github.com/rockowitz/ddcutil/issues/68 and this part about the EV2785 in the docs at  http://www.ddcutil.com/monitor_notes/ (Which is basically the same content as the issue):

> Eizo EV2785
>
> Reports EDID ad I2c bus address x.50. Does not support DDC/CI (bus address x37).
>
> Uses USB for monitor control, but uses manufacturer-defined usage pages for communication, i.e. does not support the USB Monitor Control Class Specification.

So this means, unless someone is willing and able to reverse engineer this stuff for the EV2785, there is not much hope.

But different EIZO monitors seem to work (at least partially) according to the `ddcutil` docs. For example the Coloredge CG19 seems to follow the MCCS specs (see link to monitor notes above).

**Comment by waydabber** [comment]
For those with Apple displays, there is a link to a test build with proper support in this issue: https://github.com/MonitorControl/MonitorControl/issues/539

Unfortunately this will not work for EIZO displays.

**Comment by waydabber** [comment]
I am afraid MCCS over USB probably won't happen unless somebody who actually has an EIZO display that supports this protocol and has the ability to port the relevant codes from [this](https://github.com/rockowitz/ddcutil/tree/39ce66b42305b304816d635b42a5b38ce280e2dd/src/usb) over to Swift/macOS does the job.

(Of course if somebody is willing to donate a display like this to the cause then the core dev team can work on it as well but even then it's gonna be a lot of work so I don't really see this as a feasible alternative).

I'll convert this issue into a discussion since it makes no sense to have this sit here forever while letting the stalebot eat the issue does not seem to be the proper way to handle this either.

**Reply by pjanx** [reply]
It's pointless. I have a similar piece-of-crap EIZO where the support also told me "lol no" years ago, and the best I've managed to make it do over USB is to show some kind of a message, using Windows drivers. So there's nothing useful to replay.

Newer EIZOs should support the normal protocols again, as per their specifications.

But I won't be buying EIZO anymore, I don't think so.

**Reply by gustavneustadt** [reply]
Oh man. Great display and also expensive, what a pitty. Earlier this year the tech support told me that they only use controllers without any DDC/CI since some years. 
So I think there is no improvement in sight for eizo monitors.

**Reply by pjanx** [reply]
Weird, I remember seeing it somewhere, but not anymore. Might have been a temporary mistake, CS2731/40 don't officially mention anything.

In any case, if anyone can find _any_ software to change the settings _at all_, there is hope. ColorNavigator wasn't it.

**Reply by gustavneustadt** [reply]
If you are on windows there is eizos proprietary software. But honestly this software can’t be any good ;-)

[Screen InStyle](https://www.eizo.de/zubehoer/software/screen-instyle/)

**Reply by pjanx** [reply]
It's for the EV series, as can be seen on the bottom of the page or [also here](https://www.eizoglobal.com/products/flexscan/screen-instyle/#tab04). They might not have EIZO's custom display controllers. (Update from a few years later: they do, it's just that Screen InStyle arbitrarily doesn't support ColorEdge.)

**Reply by pjanx** [reply]
A little update: apparently ColorNavigator 7 can read and set brightness, it will just be a pain to reverse engineer. (It seems that you can do quite a bit, however the libemc library that talks to displays is undocumented. The other approach is more or less blind USB packet replay.)

I've ordered a CG2700S to replace my cooked CS2730, so with some luck, I might come up with a more or less generally working reference reimplementation.

***

![cooked](https://github.com/user-attachments/assets/e66312f9-5852-49b4-b4d9-1a0423bd8945)

Thing looks like this at mere 18k hours, started failing shortly after going out of the 5 year warranty. Which is at least better than my similarly expensive BenQ, whose backlight failed shortly after going out of its 2 year warranty.

**Reply by pjanx** [reply]
While this is not really useful for this project, you can indeed reuse the libemc library that comes with ColorNavigator, in e.g. `/Library/Application Support/EIZO/ColorNavigator 7/tools/CN7ResetTool`, like this:
```c++
#include <cstdio>
#include <cstdint>

extern "C" {
	// Find all connected monitors, return their count.
	int64_t EmCtrlFind(int32_t *array, int64_t count);

	// Retrieve the serial number and product name, which are used as IDs.
	int64_t EmCtrlGetSerialProduct(
		int64_t index, char serial[9], char product[17]);

	// Get or set the brightness.
	int64_t EmCtrlGetBrightnessForPercentage(
		const char *serial, const char *product, double *arg);
	int64_t EmCtrlSetBrightnessForPercentage(
		const char *serial, const char *product, double arg);
}

int main(int argc, char *argv[]) {
	auto count = EmCtrlFind(nullptr, 0);
	printf("%d monitors found\n", (int) count);

	double brightness = 1.;
	if (argc != 2 || sscanf(argv[1], "%lf", &brightness) != 1) {
		printf("Usage: %s BRIGHTNESS\n", argv[0]);
		return 1;
	}
	for (int64_t i = 0; i < count; i++) {
		char serial[9] = {}, product[17] = {};
		if (EmCtrlGetSerialProduct(i, serial, product) < 0)
			continue;
		printf(" #%d: %s %s\n", (int) i, serial, product);

		double previous = 0.;
		EmCtrlGetBrightnessForPercentage(serial, product, &previous);
		EmCtrlSetBrightnessForPercentage(serial, product, brightness);
		printf("%f -> %f\n", previous, brightness);
	}
	return 0;
}
```
```make
bright: bright.cpp
	$(CXX) -Wall -Wextra -std=c++11 bright.cpp -L . -l emc -o $@
	install_name_tool -change @executable_path/../libs/libemc.dylib \
		@rpath/libemc.dylib $@
	install_name_tool -add_rpath @executable_path $@
clean:
	rm -f bright
```
Thus getting and setting brightness programmatically has been resolved in principle. Input switching including USB also works this way.

**Reply by pjanx** [reply]
In the end, I've created a clean implementation that even has a clicky thingy for both macOS and Windows:
https://git.janouch.name/p/usb-drivers
https://git.janouch.name/p/usb-drivers/releases

Mind you that none of it is signed, though it should just work. Take it as it is.

**Comment by helgek** [comment]
I have the EIZO 2795 WT and MonitorControl works with it for brightness and volume control. I tried ddcctl to change input but it didn't work . It get this output:

`$ ddcctl -d 1 -i 27
D: CGDisplay 69DDF3D3-639E-11A0-1A2E-F30EFCFAFBFF dispID(#1892442833) (2560x1440 0°) 109.00 DPI
I: found 1 external display
E: Failed to find display in WindowServer's preferences! (/Library/Preferences/com.apple.windowserver.plist)
I: polling EDID for #1 (ID 1892442833 => (null))
I: got edid.serial: 33829011
I: got edid.name: EV2795
D: action: i: 27
D: setting VCP control #96 => 27`

**Comment by helgek** [comment]
FYI - I just stumbled over this project: https://github.com/openeizo/openeizo

**Comment by Skirmisher** [comment]
I haven't seen this mentioned anywhere in this repo's issues, which surprises me a little bit: the early Apple Cinema/Studio Displays use (very limited) USB MCCS control, whether the USB is connected to a host USB port (DVI models) or built in to the ADC connector (broken out via the DVI-to-ADC adapter for more modern systems). Apple apparently dropped support for brightness control on these monitors circa macOS 10.9/10.10, so no facility exists to control their brightness at all, since they lack hardware controls. Under Linux, all of the Apple USB-equipped monitors through to the Thunderbolt Display can be controlled with `ddcutil`'s MCCS support (among other userspace utilities), or through the bespoke (and neglected) `appledisplay` kernel driver, which seems to implement MCCS (or at least the subset that Apple used) without realizing it.

The upshot of all this is that I can't control the brightness of my Apple Studio Display (17" LCD, ADC) under macOS (12.1.3, M1 mini), either with the system display controls or using MonitorControl. Fiddling with settings such as explicitly setting the brightness range doesn't make a difference.

**Comment by deg0nz** [comment]
Hey folks,

Suddenly my EIZO EV2785 monitors work with MonitorControl. I accidently hit the brightness buttons and my monitors adjusted their brightness o_O

I was looking for PRs regarding this here and in ddcutil but couldn't find anything about this. What happened?!

Here's my setup:
16'' MBP 2019 and 14'' MBP 2021, both running Ventura 13.1
MonitorControl v. 4.1.0 build 7034

Connections:
```
MacBook -> Sonnett Echo11 Thunderbold Dock -> Sabrent Thunderbolt Dual DisplayPort adapter -> displays via DP
                                          \-> USB3-A to USB-C upstream ports of displays
```

Edit: fix typo where I said DVI instead of DisplayPort

**Comment by gustavneustadt** [comment]
Wow. Interesting. But the EV2785 has no DVI (only USB-C DisplayPort, HDMI and DisplayPort) so I assume you connect via normal DisplayPort **and** the Monitors USB-C to USB-A?

**Reply by deg0nz** [reply]
Yeah, they are connected via DisplayPort for Video and via USB-C for using the USB hubs. 
The USB connection is actually: USB-A -> USB-A-to-C converter -> USB-C

**Reply by deg0nz** [reply]
Ahh! And I had a typo in my post, of course this is a plain DisplayPort connection!

**Reply by gustavneustadt** [reply]
But besides that — unfortunately I can't confirm this on my setup … 
I'm running Ventura 13.1.
```mermaid
graph TD
    A(Macbook Pro M1 Max) -->|Thunderbolt| B(CalDigit TS3)
    B -->|Front USB 3 to USB-C| C(EIZO EV2785)
    B -->|DisplayPort| C
```

Maybe someone else can reproduce @deg0nz approach.

**Reply by gustavneustadt** [reply]
Ah. Now I see what @deg0nz mean – I think you are seeing software dimming in action there as mentioned in the readme:

> Some displays (notably EIZO) use MCCS over USB or an entirely custom protocol for control. These displays are supported with software dimming only.

**Reply by deg0nz** [reply]
Ahh, that explains it! Thanks for clarification. It's indeed activated here.

**Comment by helgek** [comment]
Never managed to come back to the topic of analysing the USB port communication to be able to also switch monitor input signal outside MS Windows (with Screen InStyle software) but recently I discovered that EIZO has released a MacOS version of InStyle software. So for anyone who hasn't seen this yet, just wanted to let you know.
